src/output.cpp

```diff
@@ -0,0 +1,736 @@
+}
+
+std::string header(const std::string& msg) {return header(msg.c_str());}
+
+void
+header(const char* msg, int level) {
+  auto out = header(msg);
+
+  // Print header based on verbosity level.
+  if (settings::verbosity >= level)
+    std::cout << '\n' << out << "\n\n";
+}
+
+//==============================================================================
+
+std::string time_stamp()
+{
+  std::stringstream ts;
+  std::time_t t = std::time(nullptr);   // get time now
+  ts << std::put_time(std::localtime(&t), "%Y-%m-%d %H:%M:%S");
+  return ts.str();
+}
+
+//==============================================================================
+
+extern "C" void print_particle(Particle* p)
+{
+  // Display particle type and ID.
+  switch (p->type_) {
+    case Particle::Type::neutron:
+      std::cout << "Neutron ";
+      break;
+    case Particle::Type::photon:
+      std::cout << "Photon ";
+      break;
+    case Particle::Type::electron:
+      std::cout << "Electron ";
+      break;
+    case Particle::Type::positron:
+      std::cout << "Positron ";
+      break;
+    default:
+      std::cout << "Unknown Particle ";
+  }
+  std::cout << p->id_ << "\n";
+
+  // Display particle geometry hierarchy.
+  for (auto i = 0; i < p->n_coord_; i++) {
+    std::cout << "  Level " << i << "\n";
+
+    if (p->coord_[i].cell != C_NONE) {
+      const Cell& c {*model::cells[p->coord_[i].cell]};
+      std::cout << "    Cell             = " << c.id_ << "\n";
+    }
+
+    if (p->coord_[i].universe != C_NONE) {
+      const Universe& u {*model::universes[p->coord_[i].universe]};
+      std::cout << "    Universe         = " << u.id_ << "\n";
+    }
+
+    if (p->coord_[i].lattice != C_NONE) {
+      const Lattice& lat {*model::lattices[p->coord_[i].lattice]};
+      std::cout << "    Lattice          = " << lat.id_ << "\n";
+      std::cout << "    Lattice position = (" << p->coord_[i].lattice_x
+                << "," << p->coord_[i].lattice_y << ","
+                << p->coord_[i].lattice_z << ")\n";
+    }
+
+    std::cout << "    r = (" << p->coord_[i].r.x << ", "
+              << p->coord_[i].r.y << ", " << p->coord_[i].r.z << ")\n";
+    std::cout << "    u = (" << p->coord_[i].u.x << ", "
+              << p->coord_[i].u.y << ", " << p->coord_[i].u.z << ")\n";
+  }
+
+  // Display miscellaneous info.
+  if (p->surface_ != 0) {
+    const Surface& surf {*model::surfaces[std::abs(p->surface_)-1]};
+    std::cout << "  Surface = " << std::copysign(surf.id_, p->surface_) << "\n";
+  }
+  std::cout << "  Weight = " << p->wgt_ << "\n";
+  if (settings::run_CE) {
+    std::cout << "  Energy = " << p->E_ << "\n";
+  } else {
+    std::cout << "  Energy Group = " << p->g_ << "\n";
+  }
+  std::cout << "  Delayed Group = " << p->delayed_group_ << "\n";
+
+  std::cout << "\n";
+}
+
+//==============================================================================
+
+void print_plot()
+{
+  header("PLOTTING SUMMARY", 5);
+  if (settings::verbosity < 5) return;
```

Comment:
If possible, I would use a constant rather than the magic number `5` here.  Not sure what logging library / levels are being used, but typically constants are provided within the libraries.